core/src/main/java/org/infinispan/interceptors/impl/AbstractIracRemoteSiteInterceptor.java

```diff
@@ -0,0 +1,99 @@
+      final Object key = command.getKey();
+      CacheEntry<?, ?> entry = ctx.lookupEntry(key);
+      IracMetadata remoteMetadata = IracUtils.getIracMetadataFromCommand(command, key);
+      IracMetadata localMetadata = IracUtils.getIracMetadata(entry, iracVersionGenerator);
+
+      if (localMetadata == null) {
+         localMetadata = iracVersionGenerator.getTombstone(key);
+      }
+
+      assert remoteMetadata != null;
+
+      iracVersionGenerator.updateVersion(getSegment(key), remoteMetadata.getVersion());
+
+      if (localMetadata != null) {
+         validateAndSetMetadata(entry, command, localMetadata, remoteMetadata);
+      } else {
+         setIracMetadata(entry, remoteMetadata);
+      }
+   }
+
+   /**
+    * Invoked by backup owners, it make sure the entry has the same version as set by the primary owner.
+    */
+   protected void setIracMetadataForOwner(InvocationContext ctx, DataWriteCommand command,
+         @SuppressWarnings("unused") Object rv) {
+      final Object key = command.getKey();
+      IracMetadata metadata = IracUtils.getIracMetadataFromCommand(command, key);
+      assert metadata != null;
+      iracVersionGenerator.updateVersion(getSegment(key), metadata.getVersion());
+      setIracMetadata(ctx.lookupEntry(key), metadata);
+   }
+
+   private void validateAndSetMetadata(CacheEntry<?, ?> entry, DataWriteCommand command,
+         IracMetadata localMetadata, IracMetadata remoteMetadata) {
+      if (isTraceEnabled()) {
+         getLog().tracef("[IRAC] Comparing local and remote metadata: %s and %s", localMetadata, remoteMetadata);
+      }
+      IracEntryVersion localVersion = localMetadata.getVersion();
+      IracEntryVersion remoteVersion = remoteMetadata.getVersion();
+      switch (remoteVersion.compareTo(localVersion)) {
+         case CONFLICTING:
+            resolveConflict(entry, command, localMetadata, remoteMetadata);
+            return;
+         case EQUAL:
+         case BEFORE:
+            IracUtils.discardUpdate(getLog(), entry, command, remoteMetadata);
+            return;
+      }
+      setIracMetadata(entry, remoteMetadata);
+   }
+
+
+   private void resolveConflict(CacheEntry<?, ?> entry, DataWriteCommand command, IracMetadata localMetadata,
+         IracMetadata remoteMetadata) {
+      if (isTraceEnabled()) {
+         getLog().tracef("[IRAC] Conflict found between local and remote metadata: %s and %s", localMetadata,
+               remoteMetadata);
+      }
+      //same site? conflict?
+      assert !localMetadata.getSite().equals(remoteMetadata.getSite());
+      if (localMetadata.getSite().compareTo(remoteMetadata.getSite()) < 0) {
+         IracUtils.discardUpdate(getLog(), entry, command, remoteMetadata);
+         return;
+      }
+      //other site update has priority!
+      setIracMetadata(entry, remoteMetadata);
+      //TODO! this isn't required now but when we allow custom conflict resolution, we need to merge the versions!
```

Comment:
Can you create a Jira for this and add the ID to the comment.

Reply:
done: https://issues.redhat.com/browse/ISPN-11802